File: backend/dataset/maxPathSum/maxPathSum_6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode)->int:
        self.ans = -float('inf')
        
        def path(root):
            if not root:
                return 0
            left = max(path(root.left),0)
            right = max(path(root.right),0)
            val = root.val
            self.ans = max(self.ans,val+left+right)
            return val+max(left,right)
        
        path(root)
        return self.ans

File: backend/CDataset/maxPathSum/maxPathSum_6.cpp
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int ans=INT_MIN;
        dfs(root,ans);
        return ans;
    }
    int dfs(TreeNode* root, int &ans)
    {
        if(!root) return 0;
        int l=max(0,dfs(root->left,ans));
        int r=max(0,dfs(root->right,ans));      
        ans=max(ans, root->val+l+r);
        return root->val + max(l,r);
    }
};
